### PR TITLE
CMake: Do not automatically rebuild a project when switching build type

### LINKED ISSUE #7693: Change behavior of 'make debug' and 'make release'
https://github.com/geodynamics/aspect/pull/2800 is about an issue that has bugged me for a very long time already in ASPECT: When you call `make debug`, it switches to debug mode *and* re-compiles everything. This is fine if you have a tutorial with just one source file, but if you have a project with 263 source files this is annoying *because cmake compiles everything with `-j1`*. So it's exceedingly slow.

Might I suggest that we change the behavior of `make debug` and `make release` to just *switch* to the respective mode, delete the executable (to avoid anyone using it), and then asking the user to explicitly compile again?

(I'm apparently not the only one who has their own workaround to this issue -- @tjhei has as well: https://community.geodynamics.org/t/makefile-multiple-processor-issue/413?u=bangerth )

### PR DESCRIPTION
**Pull request by tamiko**
For a long time the "debug" and "release" targets of our convenience macro
DEAL_II_INVOKE_AUTOPILOT automatically rebuild the project when switching
to the debug or release flavor.

Closes #7693

### CODE REVIEW

File: cmake/macros/macro_deal_ii_invoke_autopilot.cmake [hunk 1]
```diff
@@ -141,13 +141,17 @@ MACRO(DEAL_II_INVOKE_AUTOPILOT)
   # Define custom targets to easily switch the build type:
   ADD_CUSTOM_TARGET(debug
     COMMAND ${CMAKE_COMMAND} -DCMAKE_BUILD_TYPE=Debug ${CMAKE_SOURCE_DIR}
-    COMMAND ${CMAKE_COMMAND} --build ${CMAKE_BINARY_DIR} --target all
+    COMMAND ${CMAKE_COMMAND} -E echo "***"
+    COMMAND ${CMAKE_COMMAND} -E echo "*** Switched to Release mode. Now recompile with: ${_make_command}"
```
Comment (tjhei) [inline cmake/macros/macro_deal_ii_invoke_autopilot.cmake]: I needed to add "VERBATIM" to make that work. Did you test this?

Comment (tamiko) [reply cmake/macros/macro_deal_ii_invoke_autopilot.cmake]: Wow. This indeed needs the `VERBATIM` keyword. The answer is yes I tested the functionality beforehand but not with the final three line statement but only a single-line one.

**Comment by masterleinad** [comment]
If I only compile the libraray in `Debug` mode, I get:
```
###
#
#  WARNING:
#
#  CMAKE_BUILD_TYPE "Release" unsupported by current installation!
#  deal.II was configured with "Debug".
#
#  CMAKE_BUILD_TYPE was forced to "Debug".
#
###
-- Autopilot invoked
-- Run  $ ninja info  to print a detailed help message
-- Configuring done
-- Generating done
-- Build files have been written to: /mnt/data/darndt/Sources/dealii-clang-6.0.0/examples/step-1
***
*** Switched to Release mode. Now recompile with: $ ninja
***
```

What about aborting in this case and don't proceed with the other commands?

**Comment by tjhei** [comment]
> What about aborting in this case and don't proceed with the other commands?

Or we only create the target debug if  ``${DEAL_II_BUILD_TYPE}`` contains ``"Debug"``.
Or we always provide the target and show and error ("sorry, you need to compile deal.II with DebugRelease to switch build mode")

**Comment by tamiko** [comment]
> Or we only create the target debug if ${DEAL_II_BUILD_TYPE} contains "Debug".

Good point. I went with this approach.

**Comment by tjhei** [comment]
thanks, @tamiko !

**Comment by bangerth** [comment]
By the way, would it be useful to remove the executable when the switch from debug to release or back actually affects anything? When I do
```
  make debug
  make
  make release
  ./step-xx                # runs the debug executable
```
This may be confusing.

**Pull request merged**